keps/sig-api-machinery/3945-crd-conversion-with-cel/README.md

```diff
@@ -0,0 +1,883 @@
+and validate the declared limits?
+-->
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+### Declare the conversion rules within the OpenAPIv3 schema
+
+- conversion is not logically part of a particular schema, and the schemas are used by more than just the api server
+- would have practical implications, like making schemas harder to compare or process with external systems
+
+
+### Template based conversions
+
+KCP implement CEL based version conversion by specifying a new API kind `APIConversion` to specify the conversionRules.
+
+We would prefer to define conversionRules inside CRD since it would make CRD more self-contained, and the typical use
+cases for CRD conversion would be a small subset of values needs to be converted while most of the values stay the same.
```

Comment:
I agree.  Also link to KCP implementation? It might be valuable to know what exactly they did, how similar it is to what we are doing, and how happy they are with their solution (do they have any major known feature gaps, etc...?).